profiler/src/ProfilerEngine/Datadog.Profiler.Native/ExceptionsProvider.cpp

```diff
@@ -0,0 +1,245 @@
+    _pCorProfilerInfo(pCorProfilerInfo),
+    _pManagedThreadList(pManagedThreadList),
+    _pFrameStore(pFrameStore),
+    _messageFieldOffset(),
+    _stringLengthOffset(0),
+    _stringBufferOffset(0),
+    _mscorlibModuleId(0),
+    _exceptionClassId(0)
+{
+}
+
+const char* ExceptionsProvider::GetName()
+{
+    return _serviceName;
+}
+
+bool ExceptionsProvider::OnModuleLoaded(const ModuleID moduleId)
+{
+    if (_mscorlibModuleId != 0)
+    {
+        return false;
+    }
+
+    // Check if it's mscorlib. In that case, locate the System.Exception type
+    std::string assemblyName;
+
+    if (!FrameStore::GetAssemblyName(_pCorProfilerInfo, moduleId, assemblyName))
+    {
+        Log::Warn("Failed to retrieve assembly name for module ", moduleId);
+        return false;
+    }
+
+    if (assemblyName != "System.Private.CoreLib" && assemblyName != "mscorlib")
+    {
+        return false;
+    }
+
+    _mscorlibModuleId = moduleId;
+
+    INVOKE(_pCorProfilerInfo->GetStringLayout2(&_stringLengthOffset, &_stringBufferOffset))
+
+    return true;
+}
+
+bool ExceptionsProvider::OnExceptionThrown(ObjectID thrownObjectId)
+{
+    ClassID classId;
+
+    INVOKE(_pCorProfilerInfo->GetClassFromObject(thrownObjectId, &classId))
+
+    ModuleID moduleId;
+    mdTypeDef typeDefToken;
+
+    INVOKE(_pCorProfilerInfo->GetClassIDInfo(classId, &moduleId, &typeDefToken))
+
+    ComPtr<IMetaDataImport2> metadataImport;
+
+    INVOKE(_pCorProfilerInfo->GetModuleMetaData(moduleId, ofRead, IID_IMetaDataImport2, reinterpret_cast<IUnknown**>(metadataImport.GetAddressOf())))
+
+    ULONG nameCharCount = 0;
+
+    INVOKE(metadataImport->GetTypeDefProps(typeDefToken, nullptr, 0, &nameCharCount, nullptr, nullptr))
+
+    const auto buffer = std::make_unique<WCHAR[]>(nameCharCount);
+
+    INVOKE(metadataImport->GetTypeDefProps(typeDefToken, buffer.get(), nameCharCount, &nameCharCount, nullptr, nullptr))
+
+    const auto pBuffer = buffer.get();
+
+    // Convert from UTF16 to UTF8
+    auto name = shared::ToString(shared::WSTRING(pBuffer));
+
+    if (_mscorlibModuleId == 0)
+    {
+        Log::Warn("An exception has been thrown but mscorlib is not loaded");
+        return false;
+    }
+
+    if (_exceptionClassId == 0)
+    {
+        if (!LoadExceptionMetadata())
+        {
+            return false;
+        }
+    }
+
+    const auto messageAddress = *reinterpret_cast<UINT_PTR*>(thrownObjectId + _messageFieldOffset.ulOffset);
+
+    std::string message;
+
+    if (messageAddress == 0)
+    {
+        std::cout << "Message is null" << std::endl;
+        message = std::string();
+    }
+    else
+    {
+        const auto stringLength = *reinterpret_cast<ULONG*>(messageAddress + _stringLengthOffset);
+
+        message = shared::ToString(reinterpret_cast<WCHAR*>(messageAddress + _stringBufferOffset), stringLength);
+    }
+
+    const auto collector = OsSpecificApi::CreateNewStackFramesCollectorInstance(_pCorProfilerInfo, _pManagedThreadList);
```

Comment:
maybe cache this collector and reuse it at each call . (instead of creating a new one at each call)

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/2743/commits/e4af25224bc681a7aa7a40125f5af3ccc32ad592